File: packages/runtime-go/core/pagination.go
package core

import "context"

// Page is one page of results plus whatever the envelope carried.
type Page[T any] struct {
	Items      []T
	Total      *int64
	NextCursor string
	HasMore    *bool
	// Raw is the decoded envelope, for a field besdk did not model.
	Raw any
}

// HasNextPage reports whether another page exists.
//
// Prefers what the server said. Falling back to "the page was not empty" is deliberate but weaker:
// it costs one extra request at the end of a full walk, which is far better than stopping early on
// an API that returns a short-but-not-final page.
func (p *Page[T]) HasNextPage() bool {
	if p.HasMore != nil {
		return *p.HasMore
	}
	if p.NextCursor != "" {
		return true
	}
	return len(p.Items) > 0
}

// Iterator walks pages and yields items one at a time.
//
// The Next/Current/Err shape rather than range-over-func: that needs Go 1.23, and an SDK should not
// impose a toolchain floor its users have no other reason to meet. This is also the shape Google
// Cloud's and Stripe's Go SDKs use, so it needs no learning:
//
//	it := client.Widgets.List(ctx, nil)
//	for it.Next(ctx) {
//		widget := it.Current()
//	}
//	if err := it.Err(); err != nil {
//		return err
//	}
//
// Errors are deliberately *not* returned from Next. A `for` condition that returns two values cannot
// be written, and the alternative — panicking — is wrong for a library. Err() after the loop is the
// established Go convention, matching bufio.Scanner and sql.Rows.
type Iterator[T any] struct {
	fetch   func(ctx context.Context, params map[string]any) (*Page[T], error)
	advance func(page *Page[T], params map[string]any) map[string]any

	params  map[string]any
	page    *Page[T]
	index   int
	current T
	done    bool
	err     error
}

// NewIterator builds an Iterator. Nothing is requested until the first Next, so constructing one is
// free and a caller who never loops has not paid for a fetch.
func NewIterator[T any](
	fetch func(ctx context.Context, params map[string]any) (*Page[T], error),
	advance func(page *Page[T], params map[string]any) map[string]any,
	initial map[string]any,
) *Iterator[T] {
	params := map[string]any{}
	for k, v := range initial {
		params[k] = v
	}
	return &Iterator[T]{fetch: fetch, advance: advance, params: params}
}

// Next advances to the next item, fetching a page when the current one is exhausted.
func (it *Iterator[T]) Next(ctx context.Context) bool {
	if it.err != nil || it.done {
		return false
	}
	for {
		if it.page != nil && it.index < len(it.page.Items) {
			it.current = it.page.Items[it.index]
			it.index++
			return true
		}
		if it.page != nil {
			if !it.page.HasNextPage() {
				it.done = true
				return false
			}
			next := it.advance(it.page, it.params)
			if next == nil {
				// A server that claims more but supplies no cursor is broken; looping forever on
				// its behalf would turn its bug into a hung client.
				it.done = true
				return false
			}
			it.params = next
		}
		page, err := it.fetch(ctx, it.params)
		if err != nil {
			it.err = err
			return false
		}
		it.page = page
		it.index = 0
		if len(page.Items) == 0 && !page.HasNextPage() {
			it.done = true
			return false
		}
	}
}

// Current returns the item Next just advanced to. Only valid after Next returned true.
func (it *Iterator[T]) Current() T { return it.current }

// Err returns the first error that stopped iteration, if any. Always check it after the loop.
func (it *Iterator[T]) Err() error { return it.err }

// Page returns the envelope the current item came from, for callers who need Total.
func (it *Iterator[T]) Page() *Page[T] { return it.page }

// All drains the iterator into a slice.
//
// A convenience for the common small-collection case, and deliberately not the default: an API with
// a hundred thousand widgets should not be encouraged to load them all into memory.
func (it *Iterator[T]) All(ctx context.Context) ([]T, error) {
	var out []T
	for it.Next(ctx) {
		out = append(out, it.Current())
	}
	return out, it.Err()
}

// AdvanceOffset advances an offset scheme.
//
// The caller's own parameters are copied first and the computed offset written last. That ordering is
// the whole bug the TypeScript runtime once shipped: a caller who passed offset=0 explicitly
// overwrote the computed value, so every page after the first refetched page one and the iterator
// returned duplicates forever.
func AdvanceOffset[T any](limitParam, offsetParam string) func(*Page[T], map[string]any) map[string]any {
	return func(page *Page[T], params map[string]any) map[string]any {
		current := 0
		if v, ok := params[offsetParam].(int); ok {
			current = v
		}
		step := len(page.Items)
		if v, ok := params[limitParam].(int); ok && v > 0 {
			step = v
		}
		if step <= 0 {
			return nil
		}
		next := map[string]any{}
		for k, v := range params {
			next[k] = v
		}
		next[offsetParam] = current + step
		return next
	}
}

// AdvancePageNumber advances a page-number scheme.
func AdvancePageNumber[T any](pageParam string) func(*Page[T], map[string]any) map[string]any {
	return func(page *Page[T], params map[string]any) map[string]any {
		current := 1
		if v, ok := params[pageParam].(int); ok {
			current = v
		}
		next := map[string]any{}
		for k, v := range params {
			next[k] = v
		}
		next[pageParam] = current + 1
		return next
	}
}

// AdvanceCursor advances a cursor scheme.
func AdvanceCursor[T any](cursorParam string) func(*Page[T], map[string]any) map[string]any {
	return func(page *Page[T], params map[string]any) map[string]any {
		if page.NextCursor == "" {
			return nil
		}
		next := map[string]any{}
		for k, v := range params {
			next[k] = v
		}
		next[cursorParam] = page.NextCursor
		return next
	}
}

File: packages/runtime-python/src/graft_runtime/_pagination.py
"""Pagination.

The shape here is a deliberate departure from the TypeScript runtime, and the difference is the
point: TypeScript returns one object that is both `AsyncIterable` and `PromiseLike`, so `for await`
walks items and `await` gives a page. Python has no equivalent overload, so a page **is** an
iterable of its items and `.pages()` walks envelopes:

    for widget in client.widgets.list():        # every widget, across pages
        ...

    for page in client.widgets.list().pages():  # one envelope at a time
        print(page.total)

Iterating items by default is the right default for Python because it is what every stdlib and
third-party iterable does; a caller who wanted pages asks for them.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Generic, TypeVar

if TYPE_CHECKING:
    from collections.abc import AsyncIterator, Callable, Coroutine, Iterator, Mapping, Sequence

__all__ = ["AsyncPage", "AsyncPaginator", "Page", "PageInfo", "Paginator"]

T = TypeVar("T")


class PageInfo:
    """How to request the next page, or ``None`` when there is not one."""

    __slots__ = ("params",)

    def __init__(self, params: Mapping[str, object]) -> None:
        self.params = dict(params)


class Page(Generic[T]):
    """One page of results, plus whatever the envelope carried."""

    def __init__(
        self,
        *,
        items: Sequence[T],
        total: int | None = None,
        next_cursor: str | None = None,
        has_more: bool | None = None,
        raw: object = None,
    ) -> None:
        self.items = list(items)
        self.total = total
        self.next_cursor = next_cursor
        self._has_more = has_more
        self.raw = raw

    @property
    def has_next_page(self) -> bool:
        """Whether another page exists.

        Prefers what the server said. Falling back to "the page was not empty" is deliberate but
        weaker: it costs one extra request at the end of a full walk, which is far better than
        stopping early on an API that returns a short-but-not-final page.
        """
        if self._has_more is not None:
            return self._has_more
        if self.next_cursor is not None:
            return self.next_cursor != ""
        return len(self.items) > 0

    def __iter__(self) -> Iterator[T]:
        return iter(self.items)

    def __len__(self) -> int:
        return len(self.items)

    def __bool__(self) -> bool:
        return bool(self.items)

    def __repr__(self) -> str:
        return (
            f"Page(items={len(self.items)}, total={self.total!r}, "
            f"has_next_page={self.has_next_page})"
        )


class Paginator(Generic[T]):
    """Lazily walks pages, yielding items.

    Nothing is requested until iteration begins, so building one is free and a caller who only wants
    ``.pages()`` has not already paid for a fetch.
    """

    def __init__(
        self,
        fetch: Callable[[Mapping[str, object]], Page[T]],
        *,
        initial: Mapping[str, object] | None = None,
        advance: Callable[[Page[T], Mapping[str, object]], Mapping[str, object] | None],
    ) -> None:
        self._fetch = fetch
        self._initial = dict(initial or {})
        self._advance = advance

    def pages(self) -> Iterator[Page[T]]:
        params: Mapping[str, object] | None = self._initial
        while params is not None:
            page = self._fetch(params)
            yield page
            if not page.has_next_page:
                return
            params = self._advance(page, params)

    def __iter__(self) -> Iterator[T]:
        for page in self.pages():
            yield from page.items

    def first_page(self) -> Page[T]:
        """The first page, for callers who want the envelope and nothing more."""
        return self._fetch(self._initial)


class AsyncPage(Page[T]):
    """One page from an async client. Identical envelope, so it simply reuses :class:`Page`."""

    def __aiter__(self) -> AsyncIterator[T]:
        return self._aiter()

    async def _aiter(self) -> AsyncIterator[T]:
        for item in self.items:
            yield item


class AsyncPaginator(Generic[T]):
    """The async twin of :class:`Paginator`."""

    def __init__(
        self,
        fetch: Callable[[Mapping[str, object]], Coroutine[object, object, AsyncPage[T]]],
        *,
        initial: Mapping[str, object] | None = None,
        advance: Callable[[AsyncPage[T], Mapping[str, object]], Mapping[str, object] | None],
    ) -> None:
        self._fetch = fetch
        self._initial = dict(initial or {})
        self._advance = advance

    async def pages(self) -> AsyncIterator[AsyncPage[T]]:
        params: Mapping[str, object] | None = self._initial
        while params is not None:
            page = await self._fetch(params)
            yield page
            if not page.has_next_page:
                return
            params = self._advance(page, params)

    async def __aiter__(self) -> AsyncIterator[T]:
        async for page in self.pages():
            for item in page.items:
                yield item

    async def first_page(self) -> AsyncPage[T]:
        return await self._fetch(self._initial)


def advance_offset(
    limit_param: str, offset_param: str
) -> Callable[[Page[T], Mapping[str, object]], Mapping[str, object] | None]:
    """Advance an offset scheme.

    The parameter order here is the whole bug the TypeScript runtime shipped once: the caller's
    original params must be spread **first** and the computed offset **last**, or a caller who
    passed ``offset=0`` explicitly overwrites the computed value and every page after the first
    refetches page one — an infinite iterator of duplicates. Written as an explicit dict build
    rather than a merge so the ordering cannot be reversed by accident.
    """

    def advance(page: Page[T], params: Mapping[str, object]) -> Mapping[str, object] | None:
        previous = params.get(offset_param)
        current = previous if isinstance(previous, int) else 0
        limit_value = params.get(limit_param)
        step = limit_value if isinstance(limit_value, int) and limit_value > 0 else len(page.items)
        if step <= 0:
            return None
        nxt = dict(params)
        nxt[offset_param] = current + step
        return nxt

    return advance


def advance_page_number(
    page_param: str,
) -> Callable[[Page[T], Mapping[str, object]], Mapping[str, object] | None]:
    """Advance a page-number scheme."""

    def advance(page: Page[T], params: Mapping[str, object]) -> Mapping[str, object] | None:
        previous = params.get(page_param)
        current = previous if isinstance(previous, int) else 1
        nxt = dict(params)
        nxt[page_param] = current + 1
        return nxt

    return advance


def advance_cursor(
    cursor_param: str,
) -> Callable[[Page[T], Mapping[str, object]], Mapping[str, object] | None]:
    """Advance a cursor scheme.

    Returns ``None`` on a missing cursor rather than resending the previous one: an API that says
    "there are more" but supplies no cursor is broken, and looping forever on its behalf turns a
    server bug into a hung client.
    """

    def advance(page: Page[T], params: Mapping[str, object]) -> Mapping[str, object] | None:
        if page.next_cursor is None or page.next_cursor == "":
            return None
        nxt = dict(params)
        nxt[cursor_param] = page.next_cursor
        return nxt

    return advance
